### PR TITLE
8317432: Async UL: Use memcpy instead of strcpy in Message ctr

### PR DESCRIPTION
**Pull request by jdksjolen**
Hi,

This change does three things:

1. At construction of message we used to compute `strlen` and then `strcpy`, we change this to do only one `strlen` and then a `memcpy` instead.
2. A `strlen` used to be required to calculate the size when finding the next message in the iterator. We instead store the previously computed message length and so the `strlen` is no longer necessary. The cost of this is 8 bytes extra per message (64 -> 72 bytes).
3. There's also some code which alters the alignment to explicitly handle the alignment of `Message` and ensure that `_pos` starts at a `Message`-aligned pointer.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8317432](https://bugs.openjdk.org/browse/JDK-8317432): Async UL: Use memcpy instead of strcpy in Message ctr (**Enhancement** - P4)

### Reviewers
 * [Xin Liu](https://openjdk.org/census#xliu) (@navyxliu - Committer)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16029/head:pull/16029` \
`$ git checkout pull/16029`

Update a local copy of the PR: \
`$ git checkout pull/16029` \
`$ git pull https://git.openjdk.org/jdk.git pull/16029/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16029`

View PR using the GUI difftool: \
`$ git pr show -t 16029`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16029.diff">https://git.openjdk.org/jdk/pull/16029.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16029#issuecomment-1746472454)

### CODE REVIEW

**Comment by jdksjolen** [comment]
Saving the length can be done even earlier in the UL process, both for sync and async UL. This, in turn, allows us to use `fwrite` instead of `jio_fprintf` in `logFileStreamOutput.cpp:128`. Just some small ways of making UL a bit quicker.

File: src/hotspot/share/logging/logAsyncWriter.cpp [hunk 1]
```diff
@@ -48,13 +49,14 @@ const LogDecorations& AsyncLogWriter::None = LogDecorations(LogLevel::Warning, L
                                       LogDecorators::None);
 
 bool AsyncLogWriter::Buffer::push_back(LogFileStreamOutput* output, const LogDecorations& decorations, const char* msg) {
-  const size_t sz = Message::calc_size(strlen(msg));
+  size_t len = strlen(msg);
```
Comment (shipilev) [inline src/hotspot/share/logging/logAsyncWriter.cpp:52]: Is it `const`?

**Review by shipilev — COMMENTED** [review]
Superficially okay, but we need @navyxliu to review this too.

File: src/hotspot/share/logging/logAsyncWriter.hpp [hunk 2]
```diff
@@ -66,38 +67,40 @@ class AsyncLogWriter : public NonJavaThread {
...
-  // |nullptr|_decorations|"",|pad| <- flush token
+  // |nullptr|_decorations|0|"",|pad| <- flush token
   // |<- _pos
   // ---------------------------------------------
   class Message {
     NONCOPYABLE(Message);
     ~Message() = delete;
     LogFileStreamOutput* const _output;
     const LogDecorations _decorations;
+    size_t _msglen; // Length of message excluding NUL-byte
```
[the remark this answers is not in the document]
Comment (jdksjolen) [reply src/hotspot/share/logging/logAsyncWriter.hpp:87]: Thanks, it is `const`. Same for the other comment.

**Comment by navyxliu** [comment]
hi, @jdksjolen 

I am aware of that we need to calculate strlen in the iterator. I thought that's it's one off thing so I try to leave more room for c-str. My concern is that we have a lot of one liners, so 8 bytes isn't that cheap for them. The second strlen is called in flusher thread anymore. 

I don't oppose the way your implement it. With strlen encoded, we actually can save the zero-terminator '\0' as well. we have to append it when we output c-str. it probably not worthy it. 

I like your other changes. 'alignof' resolves the undefined behavior indeed. Is that the original reason you made this change?

**Review by navyxliu — APPROVED** [review]
LGTM.

**Comment by jdksjolen** [comment]
Hi @navyxliu, thank you for taking a look at this.

> I am aware of that we need to calculate strlen in the iterator. I thought that it's one-off thing so I try to leave more room for c-str. My concern is that we have a lot of one liners, so 8 bytes are not that cheap for them. The second reason is that strlen is called in flusher thread anyway.

I understand, I was on the fence about saving or not saving the length. This is a 12.5% increase in the size of the `Message` envelope, so the envelope itself is already quite large compared to the string and I was hoping that we can avoid the `strlen` in the flusher thread in a future PR. We can still avoid the `strcpy` in the constructor and not save the size instead, that's still a good PR. Changing UL to use fat pointers can be investigated in the future.

**Comment by dholmes-ora** [comment]
So IIUC item 2 has now been dropped and so the issue title really does need to be updated?

**Comment by jdksjolen** [comment]
Hi,

We're dropping item number 2 for now in favour of a new RFE: [JDK-8317717](https://bugs.openjdk.org/browse/JDK-8317717). Number 1 and 3 will still stay. The title has been adjusted.

**Review by dholmes-ora — APPROVED** [review]
Seems reasonable.

Thanks

**Comment by jdksjolen** [comment]
/integrate

Thank you for the reviews.

**Pull request closed without merging**